### PR TITLE
Adding dynamic routing and weather service

### PR DESCRIPTION
**Pull request by saijel**
Creating new PR that was based off of #169.

List of changes:
* added structure to allow dynamic routing
* added weather microservice
* added 2 new workflows to Config.py

### CODE REVIEW

File: lucida/Makefile [hunk 1]
```diff
@@ -1,3 +1,3 @@
-SUBDIRS=commandcenter imagematching questionanswering calendar speechrecognition djinntonic
+SUBDIRS=commandcenter imagematching questionanswering calendar speechrecognition djinntonic weather
```
Comment (jhauswald) [inline lucida/Makefile]: remove `speechrecognition` so its not built by default

Comment (jhauswald) [inline lucida/Makefile]: since chrome is going to be default its painful to build

File: lucida/commandcenter/controllers/Decision.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+class Decision(object):
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, next_node=None, lucida_response=''):
+        self.next_node = next_node
+        self.lucdia_repsonse = lucida_response
+
+    @abc.abstractmethod
+    def logic_method(self, response_data, service_graph, dcm_node):
+        """Decision making logic"""
```
Comment (jhauswald) [inline lucida/commandcenter/controllers/Decision.py]: 💯

File: lucida/commandcenter/controllers/ThriftClient.py [hunk 3]
```diff
@@ -10,80 +12,132 @@
...
+        for service in Config.Service.LEARNERS['text']: # add concurrency?
+            knowledge_input = self.create_query_input(
+                text_type, [text_data], [text_id])
+            client, transport = self.get_client_transport(service)
+            log('Sending learn_text request to QA')
+            client.learn(str(LUCID),
+                self.create_query_spec('knowledge', [knowledge_input]))
+            transport.close()
+
+# TODO: I need help trying to make this method look less scary
```
Comment (jhauswald) [inline lucida/commandcenter/controllers/ThriftClient.py]: make this comment more descriptive what you would want to do (split into multiple functions, make a class) suggesting next that comes here has an idea.

File: lucida/commandcenter/controllers/dcm/IMMDCM.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import abc
+from Decision import*
+
+not_monument = 'cow'
+
+class IMMDCM(Decision):
+
```
Comment (jhauswald) [inline lucida/commandcenter/controllers/dcm/IMMDCM.py]: can you throw a doc string here describing a little what's going on. this is a template for DCM nodes so should be well documented.

Comment (jhauswald) [inline lucida/commandcenter/controllers/dcm/IMMDCM.py]: something like:
```
class IMMDCM(Decision):
"""
bla bla
"""
   def logic_bla
```

File: lucida/commandcenter/controllers/dcm/WEDCM.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import abc
+from Decision import*
+
+class WEDCM(Decision):
+
```
Comment (jhauswald) [inline lucida/commandcenter/controllers/dcm/WEDCM.py]: same here on docstring. get into the habit of docstringing.

File: lucida/commandcenter/data/class_IMM_DCM_QA_WE.txt [hunk 6]
```diff
@@ -0,0 +1,20 @@
+What's the weather here?
```
Comment (jhauswald) [inline lucida/commandcenter/data/class_IMM_DCM_QA_WE.txt]: ha really creative training data :1st_place_medal:

File: lucida/imagematching/opencv_imm/server/Image.cpp [hunk 7]
```diff
@@ -31,220 +31,223 @@ namespace po = boost::program_options;
...
-	struct timeval tp;
-	gettimeofday(&tp, NULL);
-	long int timestamp = tp.tv_sec * 1000 + tp.tv_usec / 1000;
-	string image_path = "input-" + to_string(timestamp) + ".jpg";
-	ofstream image_file(image_path, ios::binary);
-	print("Size of image written to FS: " << data.size());
-	image_file.write(data.c_str(), data.size());
-	image_file.close();
-	return image_path;
+   struct timeval tp;
```
Comment (jhauswald) [inline lucida/imagematching/opencv_imm/server/Image.cpp]: fixes for IMM DB creation right?

File: lucida/imagematching/opencv_imm/server/Image.cpp [hunk 8]
```diff
@@ -31,220 +31,223 @@ namespace po = boost::program_options;
...
 //struct timeval tv1, tv2;
 //int debug = 0;
 
 //string make_pbdesc(string img_name) {
-//	string res;
-//	size_t pos = img_name.find(".jpg");
-//	if (pos != std::string::npos) res = img_name.substr(0, pos);
-//	res = res + ".pb";
-//	return res;
+//   string res;
```
Comment (jhauswald) [inline lucida/imagematching/opencv_imm/server/Image.cpp]: what's all this commented out code? maybe not you but if you know put a comment/

File: lucida/imagematching/opencv_imm/server/Image.cpp [hunk 9]
```diff
@@ -31,220 +31,223 @@ namespace po = boost::program_options;
...
-//				img.size().height, img.channels(), img.step1());
-//		char *outTxt = tess->GetUTF8Text();
-//		// cout << "OCR Res: " << outTxt << endl;
-//	}
-//
-//	// Clean up
-//	tess->End();
+//   // TODO: add some preprocessing for a bounding box
+//
+//   tesseract::TessBaseAPI *tess = new tesseract::TessBaseAPI();
```
Comment (jhauswald) [inline lucida/imagematching/opencv_imm/server/Image.cpp]: what's all this commented code? if you don't know why its there lets remove

File: lucida/weather/WeatherConfig.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+# Weather Underground API key
+# https://www.wunderground.com/weather/api/
+WU_API_URL_BASE = 'http://api.wunderground.com/api/'
+WU_API_KEY = 'ff76e8e80c802d46' # TODO: remove after PR has been accepted # add your API key here
+
+# Open Weather Map API key
+# https://openweathermap.org/api
+OWM_API_URL_BASE = 'http://api.openweathermap.org/data/2.5/weather?'
+OWM_API_KEY = '362537b891e6df03a316e82565fe4df3' # TODO: remove after PR has been accepted # add your API key here
```
Comment (jhauswald) [inline lucida/weather/WeatherConfig.py]: Noted. have your own branch that across all the MRs applies your keys. keep rebasing and applying that branch to master.

File: lucida/imagematching/opencv_imm/server/Image.cpp (hunk 7 above)
Comment (saijel) [inline lucida/imagematching/opencv_imm/server/Image.cpp]: Yeah, the actual addition is on line 119 to convert images before pushing them to DB

File: lucida/imagematching/opencv_imm/server/Image.cpp (hunk 8 above)
Comment (saijel) [inline lucida/imagematching/opencv_imm/server/Image.cpp]: no clue why these lines are commented, will delete

**Pull request merged**